gtsam/nonlinear/GncOptimizer.h

```diff
@@ -0,0 +1,381 @@
+      }
+      return result;
+    }
+
+    for (size_t iter = 0; iter < params_.maxIterations; iter++) {
+
+      // display info
+      if (params_.verbosityGNC >= GncParameters::VerbosityGNC::VALUES) {
+        result.print("result\n");
+        std::cout << "mu: " << mu << std::endl;
+        std::cout << "weights: " << weights_ << std::endl;
+      }
+      // weights update
+      weights_ = calculateWeights(result, mu);
+
+      // variable/values update
+      NonlinearFactorGraph graph_iter = this->makeWeightedGraph(weights_);
+      GaussNewtonOptimizer baseOptimizer_iter(graph_iter, state_);
+      result = baseOptimizer_iter.optimize();
+
+      // update mu
+      mu_prev = mu;
+      mu = updateMu(mu);
+
+      // stopping condition
+      if (checkMuConvergence(mu, mu_prev)) {
+        // display info
+        if (params_.verbosityGNC >= GncParameters::VerbosityGNC::SUMMARY) {
+          std::cout << "final iterations: " << iter << std::endl;
+          std::cout << "final mu: " << mu << std::endl;
+          std::cout << "final weights: " << weights_ << std::endl;
+        }
+        break;
+      }
+    }
+    return result;
+  }
+
+  /// initialize the gnc parameter mu such that loss is approximately convex (remark 5 in GNC paper)
+  double initializeMu() const {
+    // compute largest error across all factors
+    double rmax_sq = 0.0;
+    for (size_t i = 0; i < nfg_.size(); i++) {
+      if (nfg_[i]) {
+        rmax_sq = std::max(rmax_sq, nfg_[i]->error(state_));
+      }
+    }
+    // set initial mu
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return 2 * rmax_sq / params_.barcSq; // initial mu
+    case GncParameters::TLS:
+      // initialize mu to the value specified in Remark 5 in GNC paper
+      // degenerate case: residual is close to zero so mu approximately equals
+      // to -1
+      return params_.barcSq / (2 * rmax_sq - params_.barcSq);
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::initializeMu: called with unknown loss type.");
+    }
+  }
+
+  /// update the gnc parameter mu to gradually increase nonconvexity
+  double updateMu(const double mu) const {
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return std::max(1.0, mu / params_.muStep); // reduce mu, but saturate at 1
+    case GncParameters::TLS:
+      // increases mu at each iteration
+      return mu * params_.muStep;
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::updateMu: called with unknown loss type.");
+    }
+  }
+
+  /// check if we have reached the value of mu for which the surrogate loss matches the original loss
+  bool checkMuConvergence(const double mu, const double mu_prev) const {
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return std::fabs(mu - 1.0) < 1e-9; // mu=1 recovers the original GM function
+    case GncParameters::TLS:
+      return std::fabs(mu - mu_prev) < params_.relativeMuTol;
```

Comment:
@jingnanshi : this does not make much sense, since the decrease in mu is constant. maybe @pantonante has a better check?

Reply:
@lucacarlone : good catch! this should be cost - cost_prev

Reply:
I think Pasquale checks if the weights are (close to) binary, which is the case after TLS converges.

Reply:
@lucacarlone : I just checked Pasquale's email, and he suggested using the relative change of the weighted sum of the squared residuals, which is what I used in TEASER++. I think I'll implement a checkConvergence function that checks mu or relative cost change depending on loss type.

Reply:
Yes, I suggest to use the relative change of the weighted sum of the squared residuals because it works well in practice. Here seems you are checking that they are not changing much, this doesn't work for TLS.

Reply:
@jingnanshi : I would implement this in the main loop (before "if (checkMuConvergence(mu, mu_prev))"): I would simply add "if (checkCostConvergence(..)  OR checkWeightsConvergence(..))" and implement these in a separate function. Let me know if you want me to take a stab - I can try to do that tonight

Reply:
@lucacarlone : That makes sense. I won't be able to do it tonight unfortunately so feel free to take a look if you have time!

Reply:
@jingnanshi @pantonante : I fixed the stopping conditions and a few other things (including setting TLS to default) - please review carefully and let me know if you have more comments.

Reply:
@lucacarlone : Had two very minor fixes. Everything looks good!

Reply:
perfect - thanks @jingnanshi !